### PR TITLE
Improve the completer module.

### PR DESCRIPTION
**Pull request by knightburton**
The monaco intelliSense will be extended with the iotjs function list
based on the manually required modules. The process, events and timers
modules are required by default.
The improve: the completer module will store the manually required
module's variable name, too. If the user hit the dot (.) character,
the completer will trigger and will try to find the proper module
function list based on the text (aka. variable name) before the dot character.

IoT.jsCode-DCO-1.0-Signed-off-by: Imre Kiss kissi.szeged@partner.samsung.com

### CODE REVIEW

File: src/app/completer.js [hunk 1]
```diff
@@ -25,8 +25,51 @@ export default class Completer {
...
+        // Get editor content before the pointer.
+        const textUntilPosition = model.getValueInRange({
+          startLineNumber: 1,
+          startColumn: 1,
+          endLineNumber: position.lineNumber,
+          endColumn: position.column,
+        });
+
+        // These modules can be used without require in IoT.js.
+        const defaultModules = [{
```
Comment (daeyeon) [inline src/app/completer.js]: It looks proper to get this const variable out of here.

**Comment by daeyeon** [comment]
I hope others also review the part related to collecting completion proposals.

File: src/app/completer.js (hunk 1 above)
Comment (knightburton) [reply src/app/completer.js]: Indeed, you are right.

File: src/app/completer.js [hunk 2]
```diff
@@ -39,41 +84,28 @@ export default class Completer {
...
-        name = name[name.length - 1];
-
-        if (!modules.includes(name)) {
-          modules.push(name.toLowerCase());
-        }
-      }
-    }
+    // Return the collected required module names and variable names.
+    return source.split('\n').filter(line => rm.test(line)).map(m => {
+      const info = m.replace(/\n/g, '').replace(/\s/g, '').split('=');
```
Comment (rtakacs) [inline src/app/completer.js]: You can merge the replacer functions:
```
m.replace(/[\n\s]/g, '').split('=');
m.replace(/\n|\s/g, '').split('=');
```

File: src/app/completer.js [hunk 3]
```diff
@@ -39,41 +84,28 @@ export default class Completer {
...
 
-    return modules;
+      return {
+        link: info.shift().replace(/(var|let|const)/g, ''),
+        mod: info.shift().replace(/.*require\s*\(/g, '').replace(/\s*('|")/g, '').replace(/\s*\);?/g, ''),
```
Comment (rtakacs) [inline src/app/completer.js]: What about using regexp groups?

In this case it doesn't need  to use the replacer function multiple times (to remove the unnecessary sub-strings). It's simply enough to read the value of the appropriate group.
In this case the `info` variable is unnecessary, just the `rm` variable should be extended with groups.

**Comment by knightburton** [comment]
@rtakacs Thank you for the review, I've updated the PR.

**Comment by rtakacs** [comment]
LGTM

**Comment by yichoi** [comment]
LGTM

**Pull request merged**